skiko/src/jvmMain/cpp/windows/directXRedrawer.cc

```diff
@@ -0,0 +1,343 @@
+              LPCSTR                 pEntrypoint,
+              LPCSTR                 pTarget,
+              UINT                   Flags1,
+              UINT                   Flags2,
+              ID3DBlob               **ppCode,
+              ID3DBlob               **ppErrorMsgs
+    );
+    static D3DCompile_t impl = nullptr;
+    if (!impl) {
+        auto d3dcompilerdll = LoadLibrary(TEXT("d3dcompiler_47.dll"));
+        if (!d3dcompilerdll)
+            return E_NOTIMPL;
+        impl = (D3DCompile_t)GetProcAddress(d3dcompilerdll, "D3DCompile");
+        if (!impl)
+            return E_NOTIMPL;
+    }
+    return impl(pSrcData, SrcDataSize, pSourceName, pDefines, pInclude, pEntrypoint, pTarget, Flags1, Flags2, ppCode, ppErrorMsgs);
+}
+
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_makeDirectXContext(
+        JNIEnv* env, jobject redrawer, jlong devicePtr)
+    {
+        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice*>(devicePtr);
+        GrD3DBackendContext backendContext = d3dDevice->backendContext;
+        return toJavaPointer(GrDirectContext::MakeDirect3D(backendContext).release());
+    }
+
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_makeDirectXRenderTarget(
+        JNIEnv * env, jobject redrawer, jlong devicePtr, jint width, jint height)
+    {
+        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice*>(devicePtr);
+
+        if (!d3dDevice->isSizeEqualTo(width, height))
+        {
+            GR_D3D_CALL_ERRCHECK(d3dDevice->swapChain->ResizeBuffers(BuffersCount, width, height, DXGI_FORMAT_R8G8B8A8_UNORM, 0));
+            d3dDevice->bufferWidth = width;
+            d3dDevice->bufferHeight = height;
+        }
+
+        d3dDevice->bufferIndex = d3dDevice->swapChain->GetCurrentBackBufferIndex();
+        ID3D12Resource* buffer;
+        GR_D3D_CALL_ERRCHECK(d3dDevice->swapChain->GetBuffer(d3dDevice->bufferIndex, IID_PPV_ARGS(&buffer)));
+        GrD3DTextureResourceInfo info(buffer,
+                                      nullptr,
+                                      D3D12_RESOURCE_STATE_PRESENT,
+                                      DXGI_FORMAT_R8G8B8A8_UNORM,
+                                      1,
+                                      1,
+                                      0);
+        GrBackendRenderTarget* renderTarget = new GrBackendRenderTarget(width, height, info);
+        return toJavaPointer(renderTarget);
+    }
+
+    void defineHardwareAdapter(IDXGIFactory4 *pFactory, IDXGIAdapter1 **ppAdapter)
+    {
+        *ppAdapter = nullptr;
+        for (UINT adapterIndex = 0;; ++adapterIndex)
+        {
+            IDXGIAdapter1 *pAdapter = nullptr;
+            if (DXGI_ERROR_NOT_FOUND == pFactory->EnumAdapters1(adapterIndex, &pAdapter))
+            {
+                break;
+            }
+            if (SUCCEEDED(D3D12CreateDevice(pAdapter, D3D_FEATURE_LEVEL_11_0, _uuidof(ID3D12Device), nullptr)))
```

Comment:
Shouldn't we use `GR_D3D_CALL_ERRCHECK` everywhere?

Reply:
We use `GR_D3D_CALL_ERRCHECK` almost everywhere, but if we need a result, we use` SUCCESS`.

Reply:
Do we need to return 0 here and return 0 in `createDirectXDevice` if `defineHardwareAdapter` failed?

Reply:
Added adapter check.